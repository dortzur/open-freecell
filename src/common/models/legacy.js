import _times from 'lodash/times';
import { srand } from '../utils';

export const SUITS = ['C', 'D', 'H', 'S'];
export const RANKS = [
  'A',
  '2',
  '3',
  '4',
  '5',
  '6',
  '7',
  '8',
  '9',
  'T',
  'J',
  'Q',
  'K',
];
export const CELL_TYPES = {
  HOME_CELL: 'HOME_CELL',
  FREE_CELL: 'FREE_CELL',
  COLUMN_CELL: 'COLUMN_CELL',
};
const getCardValue = (rank) => {
  switch (rank) {
    case 'A':
      return 1;
    case 'T':
      return 10;
    case 'J':
      return 11;
    case 'Q':
      return 12;
    case 'K':
      return 13;
    default:
      return Number.parseInt(rank);
  }
};
const getCardColor = (suit) => {
  if (suit === 'D' || suit === 'H') return 'red';
  return 'black';
};
export const createCard = (notation) => {
  const suit = notation[1];
  const rank = notation[0];
  const value = getCardValue(rank);
  const color = getCardColor(suit);

  return {
    suit,
    rank,
    value,
    color,
    notation,
  };
};

export const createDeck = () => {
  const deck = [];
  for (let i = 0; i < 13; i++) {
    for (let j = 0; j < 4; j++) {
      const card = createCard(RANKS[i] + SUITS[j]);
      deck.push(card);
    }
  }
  return deck;
};

export const createGameDeck = (gameNumber) => {
  let deck = createDeck();
  const rand = srand(gameNumber);
  for (let i = deck.length - 1; i > 0; i--) {
    const r = rand() % (i + 1);
    const card = deck[r];
    deck[r] = deck[i];
    deck[i] = card;
  }
  return deck.reverse();
};

const createCell = (cellType, index) => ({ type: cellType, index, stack: [] });
export const createGame = (gameNumber) => {
  const deck = createGameDeck(gameNumber);
  const FCS = _times(4, (index) => createCell(CELL_TYPES.FREE_CELL, index));
  const HCS = _times(4, (index) => createCell(CELL_TYPES.HOME_CELL, index));
  const CCS = _times(8, (index) => createCell(CELL_TYPES.COLUMN_CELL, index));

  for (let i = 0; i < deck.length; i++) {
    CCS[i % 8].stack.push(deck[i]);
  }
  const game = FCS.concat(HCS).concat(CCS);
  return game;
};
