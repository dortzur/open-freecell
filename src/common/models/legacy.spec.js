import { createGame,createDeck } from './legacy';

describe('legacy', () => {
  it('creates game 100', () => {

     const baba=createDeck();

  });
});
