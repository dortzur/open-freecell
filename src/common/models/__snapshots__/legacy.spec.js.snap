// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`legacy creates game 100 1`] = `
Array [
  Object {
    "index": 0,
    "stack": Array [],
    "type": "FREE_CELL",
  },
  Object {
    "index": 1,
    "stack": Array [],
    "type": "FREE_CELL",
  },
  Object {
    "index": 2,
    "stack": Array [],
    "type": "FREE_CELL",
  },
  Object {
    "index": 3,
    "stack": Array [],
    "type": "FREE_CELL",
  },
  Object {
    "index": 0,
    "stack": Array [],
    "type": "HOME_CELL",
  },
  Object {
    "index": 1,
    "stack": Array [],
    "type": "HOME_CELL",
  },
  Object {
    "index": 2,
    "stack": Array [],
    "type": "HOME_CELL",
  },
  Object {
    "index": 3,
    "stack": Array [],
    "type": "HOME_CELL",
  },
  Object {
    "index": 0,
    "stack": Array [
      Object {
        "color": "red",
        "notation": "AD",
        "rank": "A",
        "suit": "D",
        "value": 1,
      },
      Object {
        "color": "red",
        "notation": "QH",
        "rank": "Q",
        "suit": "H",
        "value": 12,
      },
      Object {
        "color": "black",
        "notation": "QS",
        "rank": "Q",
        "suit": "S",
        "value": 12,
      },
      Object {
        "color": "black",
        "notation": "6S",
        "rank": "6",
        "suit": "S",
        "value": 6,
      },
      Object {
        "color": "black",
        "notation": "3C",
        "rank": "3",
        "suit": "C",
        "value": 3,
      },
      Object {
        "color": "black",
        "notation": "2C",
        "rank": "2",
        "suit": "C",
        "value": 2,
      },
      Object {
        "color": "black",
        "notation": "8S",
        "rank": "8",
        "suit": "S",
        "value": 8,
      },
    ],
    "type": "COLUMN_CELL",
  },
  Object {
    "index": 1,
    "stack": Array [
      Object {
        "color": "black",
        "notation": "JS",
        "rank": "J",
        "suit": "S",
        "value": 11,
      },
      Object {
        "color": "black",
        "notation": "TS",
        "rank": "T",
        "suit": "S",
        "value": 10,
      },
      Object {
        "color": "red",
        "notation": "TH",
        "rank": "T",
        "suit": "H",
        "value": 10,
      },
      Object {
        "color": "red",
        "notation": "5H",
        "rank": "5",
        "suit": "H",
        "value": 5,
      },
      Object {
        "color": "black",
        "notation": "KS",
        "rank": "K",
        "suit": "S",
        "value": 13,
      },
      Object {
        "color": "red",
        "notation": "2H",
        "rank": "2",
        "suit": "H",
        "value": 2,
      },
      Object {
        "color": "black",
        "notation": "KC",
        "rank": "K",
        "suit": "C",
        "value": 13,
      },
    ],
    "type": "COLUMN_CELL",
  },
  Object {
    "index": 2,
    "stack": Array [
      Object {
        "color": "black",
        "notation": "4S",
        "rank": "4",
        "suit": "S",
        "value": 4,
      },
      Object {
        "color": "black",
        "notation": "5C",
        "rank": "5",
        "suit": "C",
        "value": 5,
      },
      Object {
        "color": "black",
        "notation": "7S",
        "rank": "7",
        "suit": "S",
        "value": 7,
      },
      Object {
        "color": "black",
        "notation": "6C",
        "rank": "6",
        "suit": "C",
        "value": 6,
      },
      Object {
        "color": "black",
        "notation": "JC",
        "rank": "J",
        "suit": "C",
        "value": 11,
      },
      Object {
        "color": "red",
        "notation": "KH",
        "rank": "K",
        "suit": "H",
        "value": 13,
      },
      Object {
        "color": "black",
        "notation": "9S",
        "rank": "9",
        "suit": "S",
        "value": 9,
      },
    ],
    "type": "COLUMN_CELL",
  },
  Object {
    "index": 3,
    "stack": Array [
      Object {
        "color": "black",
        "notation": "QC",
        "rank": "Q",
        "suit": "C",
        "value": 12,
      },
      Object {
        "color": "red",
        "notation": "9D",
        "rank": "9",
        "suit": "D",
        "value": 9,
      },
      Object {
        "color": "red",
        "notation": "8H",
        "rank": "8",
        "suit": "H",
        "value": 8,
      },
      Object {
        "color": "red",
        "notation": "3D",
        "rank": "3",
        "suit": "D",
        "value": 3,
      },
      Object {
        "color": "black",
        "notation": "3S",
        "rank": "3",
        "suit": "S",
        "value": 3,
      },
      Object {
        "color": "black",
        "notation": "8C",
        "rank": "8",
        "suit": "C",
        "value": 8,
      },
      Object {
        "color": "black",
        "notation": "5S",
        "rank": "5",
        "suit": "S",
        "value": 5,
      },
    ],
    "type": "COLUMN_CELL",
  },
  Object {
    "index": 4,
    "stack": Array [
      Object {
        "color": "black",
        "notation": "7C",
        "rank": "7",
        "suit": "C",
        "value": 7,
      },
      Object {
        "color": "red",
        "notation": "4H",
        "rank": "4",
        "suit": "H",
        "value": 4,
      },
      Object {
        "color": "black",
        "notation": "9C",
        "rank": "9",
        "suit": "C",
        "value": 9,
      },
      Object {
        "color": "red",
        "notation": "4D",
        "rank": "4",
        "suit": "D",
        "value": 4,
      },
      Object {
        "color": "black",
        "notation": "TC",
        "rank": "T",
        "suit": "C",
        "value": 10,
      },
      Object {
        "color": "red",
        "notation": "7D",
        "rank": "7",
        "suit": "D",
        "value": 7,
      },
    ],
    "type": "COLUMN_CELL",
  },
  Object {
    "index": 5,
    "stack": Array [
      Object {
        "color": "red",
        "notation": "6D",
        "rank": "6",
        "suit": "D",
        "value": 6,
      },
      Object {
        "color": "red",
        "notation": "JD",
        "rank": "J",
        "suit": "D",
        "value": 11,
      },
      Object {
        "color": "red",
        "notation": "6H",
        "rank": "6",
        "suit": "H",
        "value": 6,
      },
      Object {
        "color": "black",
        "notation": "2S",
        "rank": "2",
        "suit": "S",
        "value": 2,
      },
      Object {
        "color": "black",
        "notation": "4C",
        "rank": "4",
        "suit": "C",
        "value": 4,
      },
      Object {
        "color": "red",
        "notation": "8D",
        "rank": "8",
        "suit": "D",
        "value": 8,
      },
    ],
    "type": "COLUMN_CELL",
  },
  Object {
    "index": 6,
    "stack": Array [
      Object {
        "color": "red",
        "notation": "JH",
        "rank": "J",
        "suit": "H",
        "value": 11,
      },
      Object {
        "color": "red",
        "notation": "KD",
        "rank": "K",
        "suit": "D",
        "value": 13,
      },
      Object {
        "color": "black",
        "notation": "AC",
        "rank": "A",
        "suit": "C",
        "value": 1,
      },
      Object {
        "color": "red",
        "notation": "9H",
        "rank": "9",
        "suit": "H",
        "value": 9,
      },
      Object {
        "color": "red",
        "notation": "5D",
        "rank": "5",
        "suit": "D",
        "value": 5,
      },
      Object {
        "color": "red",
        "notation": "QD",
        "rank": "Q",
        "suit": "D",
        "value": 12,
      },
    ],
    "type": "COLUMN_CELL",
  },
  Object {
    "index": 7,
    "stack": Array [
      Object {
        "color": "red",
        "notation": "TD",
        "rank": "T",
        "suit": "D",
        "value": 10,
      },
      Object {
        "color": "red",
        "notation": "AH",
        "rank": "A",
        "suit": "H",
        "value": 1,
      },
      Object {
        "color": "black",
        "notation": "AS",
        "rank": "A",
        "suit": "S",
        "value": 1,
      },
      Object {
        "color": "red",
        "notation": "7H",
        "rank": "7",
        "suit": "H",
        "value": 7,
      },
      Object {
        "color": "red",
        "notation": "3H",
        "rank": "3",
        "suit": "H",
        "value": 3,
      },
      Object {
        "color": "red",
        "notation": "2D",
        "rank": "2",
        "suit": "D",
        "value": 2,
      },
    ],
    "type": "COLUMN_CELL",
  },
]
`;
